jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java

```diff
@@ -1,17 +1,111 @@
+    public static final int MAXIMUM_QUERY_FOR_OBJECT_RESULT = 1;
 
     private final DataSource dataSource;
 
     public JdbcTemplate(final DataSource dataSource) {
         this.dataSource = dataSource;
     }
+
+    public void update(final String sql, Object... args) {
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)
+        ) {
+            log.debug("query : {}", sql);
+
+            setPreparedStatement(pstmt, args);
+            pstmt.executeUpdate();
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+            throw new DataAccessException(e);
+        }
+    }
+
+    private void setPreparedStatement(final PreparedStatement pstmt, final Object[] args) throws SQLException {
+        for (int i = EMPTY_COUNT; i < args.length; i++) {
+            pstmt.setObject(i + MAXIMUM_QUERY_FOR_OBJECT_RESULT, args[i]);
+        }
+    }
+
+    public <T> List<T> query(final String sql, final RowMapper<T> rowMapper) {
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql);
+             ResultSet rs = pstmt.executeQuery()
+        ) {
+            log.debug("query : {}", sql);
+
+            return extractData(rowMapper, rs);
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+            throw new DataAccessException(e);
+        }
+    }
+
+    public <T> List<T> query(final String sql, final RowMapper<T> rowMapper, final Object... args) {
+        ResultSet rs = null;
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)
+        ) {
+            log.debug("query : {}", sql);
+            setPreparedStatement(pstmt, args);
+            rs = pstmt.executeQuery();
+
+            return extractData(rowMapper, rs);
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+            throw new DataAccessException(e);
+        } finally {
+            closeResultSet(rs);
+        }
+    }
+
+    private <T> List<T> extractData(final RowMapper<T> rowMapper, final ResultSet rs) throws SQLException {
+        List<T> results = new ArrayList<>();
+        while (rs.next()) {
+            results.add(rowMapper.mapRow(rs, rs.getRow()));
+        }
+        return results;
+    }
+
+    public <T> T queryForObject(final String sql, final RowMapper<T> rowMapper, Object... args) {
+        List<T> results = query(sql, rowMapper, args);
+        return getSingleResult(results);
+    }
+
+    private <T> T getSingleResult(final List<T> results) {
+        if (results.isEmpty()) {
+            throw new DataAccessException(QUERY_RESULT_EXCEPTION_MESSAGE + EMPTY_COUNT);
+        }
+        if (results.size() > MAXIMUM_QUERY_FOR_OBJECT_RESULT) {
+            throw new DataAccessException(QUERY_RESULT_EXCEPTION_MESSAGE + results.size());
+        }
```

Comment:
```suggestion
        if (results.size() != MAXIMUM_QUERY_FOR_OBJECT_RESULT) {
            throw new DataAccessException(QUERY_RESULT_EXCEPTION_MESSAGE + results.size());
        }
```

이런 식으로도 가능할 것 같은데 굳이 분리하신 이유가 있을까요?

Reply:
아무 생각 없이 조회 결과가 없을 때도 예외가 발생해야 하고
2개 이상일 때도 예외가 발생해야 한다고 생각해서 이렇게 구현한 것 같습니다. 고쳐보겠습니다!